pkg/tecdsa/signature_test.go

```diff
@@ -0,0 +1,182 @@
+package tecdsa
+
+import (
+	"github.com/keep-network/keep-core/pkg/internal/testutils"
+	"math/big"
+	"testing"
+)
+
+func TestSignatureEquals(t *testing.T) {
+	var tests = map[string]struct {
+		signature1     *Signature
+		signature2     *Signature
+		expectedResult bool
+	}{
+		"both signatures are nil": {
+			signature1:     nil,
+			signature2:     nil,
+			expectedResult: true,
+		},
+		"first signature is nil": {
+			signature1: nil,
+			signature2: &Signature{
+				R:          nil,
+				S:          nil,
+				RecoveryID: 0,
+			},
+			expectedResult: false,
+		},
+		"second signature is nil": {
+			signature1: &Signature{
+				R:          nil,
+				S:          nil,
+				RecoveryID: 0,
+			},
+			signature2:     nil,
+			expectedResult: false,
+		},
+		"both signatures' fields are non-nil and same": {
+			signature1: &Signature{
+				R:          big.NewInt(100),
+				S:          big.NewInt(200),
+				RecoveryID: 2,
+			},
+			signature2: &Signature{
+				R:          big.NewInt(100),
+				S:          big.NewInt(200),
+				RecoveryID: 2,
+			},
+			expectedResult: true,
+		},
+		"both signatures' R are nil": {
+			signature1: &Signature{
+				R:          nil,
+				S:          big.NewInt(200),
+				RecoveryID: 2,
+			},
+			signature2: &Signature{
+				R:          nil,
+				S:          big.NewInt(200),
+				RecoveryID: 2,
+			},
+			expectedResult: true,
+		},
```

Comment:
I would also add two scenarios:
- `both signatures' R are nil but recovery IDs are different`
- `both signatures' S are nil but recovery IDs are different`

Essentially testing that in case of `if s.R == nil || other.R == nil` we test only `if s.R != other.R` instead of staying inside that if and comparing only `s.R` against each other.

Reply:
Good catch! See bbc57c02